File: tsc/src/add-node.test.ts
import { describe, expect, test } from "bun:test";
import { addNode } from "./add-node";

describe("addNode", () => {
  test("should create a new node if root is null", () => {
    const root = null;

    const result = addNode(root, 5);

    expect(result).toEqual({ value: 5, left: null, right: null });
  });

  test("should add a node to left one level deep", () => {
    const root = {
      value: 10,
      left: null,
      right: null,
    };

    const result = addNode(root, 5);

    expect(result).toEqual({
      value: 10,
      left: {
        value: 5,
        left: null,
        right: null,
      },
      right: null,
    });
  });

  test("should add a node to right one level deep", () => {
    const root = {
      value: 10,
      left: null,
      right: null,
    };

    const result = addNode(root, 11);

    expect(result).toEqual({
      value: 10,
      left: null,
      right: {
        value: 11,
        left: null,
        right: null,
      },
    });
  });

  test("should add a node to right if values are equal", () => {
    const root = {
      value: 10,
      left: null,
      right: null,
    };

    const result = addNode(root, 10);

    expect(result).toEqual({
      value: 10,
      left: null,
      right: {
        value: 10,
        left: null,
        right: null,
      },
    });
  });

  test("should add a node to correct place for a deeper tree", () => {
    const root = {
      value: 10,
      left: {
        value: 5,
        left: {
          value: 2,
          left: null,
          right: null,
        },
        right: null,
      },
      right: {
        value: 15,
        left: null,
        right: null,
      },
    };

    const result = addNode(root, 7);

    expect(result).toEqual({
      value: 10,
      left: {
        value: 5,
        left: {
          value: 2,
          left: null,
          right: null,
        },
        right: {
          value: 7,
          left: null,
          right: null,
        },
      },
      right: {
        value: 15,
        left: null,
        right: null,
      },
    });
  });
});

File: go/add-node_test.go
package solution

import (
	"reflect"
	"testing"
)

type addNodeTest struct {
	title    string
	root     *Node
	value    int
	expected *Node
}

var addNodeTests = []addNodeTest{
	{"creates new root if root is nill", nil, 4, &Node{4, nil, nil}},
	{"adds value to left", &Node{10, nil, nil}, 4, &Node{10, &Node{4, nil, nil}, nil}},
	{"adds value to right", &Node{10, nil, nil}, 11, &Node{10, nil, &Node{11, nil, nil}}},
	{"adds value to right if it is equal", &Node{10, nil, nil}, 10, &Node{10, nil, &Node{10, nil, nil}}},
	{"adds value to complicated position in the tree", &Node{10, &Node{4, &Node{2, nil, nil}, nil}, &Node{11, nil, nil}}, 3, &Node{10, &Node{4, &Node{2, nil, &Node{3, nil, nil}}, nil}, &Node{11, nil, nil}}},
}

func TestAddNode(t *testing.T) {
	for _, test := range addNodeTests {
		result := addNode(test.root, test.value)

		if !reflect.DeepEqual(result, test.expected) {
			t.Error("addNode - "+test.title+",expected:", test.expected, ",got:", result)
		} else {
			t.Log("addNode - " + test.title)
		}
	}
}
